### PR TITLE
Add support GitHub Actions

### PR DESCRIPTION
**Pull request by Mephistophiles**
This PR introduce:

- fix clippy errors
- added GitHub Actions support:
  - Every push check:
    - run cargo-test (compile test and launch test suits)
    - run cargo-clippy (check additional lints)
    - run cargo-audit (check for rust vulnerabilities)
    - run cargo-fmt (check code format)
  - Every PR check

The jobs use an ubuntu-latest image with an archlinux docker inside.

P.S. cargo clippy --all-features failed (alpm package failed)

TODO:

- [ ] Remove Github tokens from PR
- [ ] Add tests for git features (by example https://github.com/Morganamilo/alpm.rs/blob/master/.github/workflows/alpm.yml)

### CODE REVIEW

**Comment by Morganamilo** [comment]
The code changes are probably fine. The github actions stuff though I'll need some explanation of some of the bits.

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -0,0 +1,99 @@
+name: Rust
+
+on:
+  pull_request:
+  push:
+    branches:
+    - master
```
Comment (Morganamilo) [inline .github/workflows/rust.yml:7]: Id prefer this on every branch. I don't always work on master.

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      with:
+        token: ${{ secrets.GITHUB_TOKEN }}
+
+  cargo-test:
+    name: Cargo Test
+    runs-on: ubuntu-latest
+
+    container:
+      image: archlinux:base-devel
+      options: --privileged
```
Comment (Morganamilo) [inline .github/workflows/rust.yml:33]: What does this do and why is it needed?

File: .github/workflows/rust.yml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      volumes:
+        - /sys/fs/cgroup:/sys/fs/cgroup
+
+    steps:
+    - name: Checkout repository
+      uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
+
+    - name: Install Rust toolchain
```
Comment (Morganamilo) [inline .github/workflows/rust.yml:43]: I'd prefer to just use pacman here. Less complex and has the added benefit of also testing paru compiles with the version of rust in the repos.

File: .github/workflows/rust.yml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    runs-on: ubuntu-latest
+
+    container:
+      image: archlinux:base-devel
+      options: --privileged
+      volumes:
+        - /sys/fs/cgroup:/sys/fs/cgroup
+
+    steps:
+    - name: Install libclang (needed by clippy)
```
Comment (Morganamilo) [inline .github/workflows/rust.yml:85]: clippy does not need clang afaik where did you get that from?

File: .github/workflows/rust.yml [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+      run: sudo pacman --noconfirm -Syy clang
```
Comment (Morganamilo) [inline .github/workflows/rust.yml:86]: Please don't -Sy or -Syy.

File: .github/workflows/rust.yml [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    - uses: actions-rs/toolchain@v1
+      with:
+          toolchain: nightly
+          components: clippy
+          override: true
+    - uses: actions-rs/clippy-check@v1
+      with:
+        token: ${{ secrets.GITHUB_TOKEN }}
+        args: #--all-features (TODO: fix all features build)
```
Comment (Morganamilo) [inline .github/workflows/rust.yml:99]: Don't just blindly test all features. There needs to be a separate test for git and non git builds.

File: .github/workflows/rust.yml (hunk 1 above)
Comment (Mephistophiles) [reply .github/workflows/rust.yml]: Ok, I'll fix that.

File: .github/workflows/rust.yml (hunk 5 above)
Comment (Mephistophiles) [reply .github/workflows/rust.yml:79]: The cache has been removed from the archive images archive.

```
warning: database file for 'core' does not exist (use '-Sy' to download)
warning: database file for 'extra' does not exist (use '-Sy' to download)
warning: database file for 'community' does not exist (use '-Sy' to download)
```

Comment (Morganamilo) [reply .github/workflows/rust.yml:79]: Then use -Syu to avoid a partial upgrade.

Comment (Mephistophiles) [reply .github/workflows/rust.yml]: Hmm, After installing rust from the official repository, the clang was not required.

File: .github/workflows/rust.yml (hunk 6 above)
Comment (Mephistophiles) [reply .github/workflows/rust.yml]: OK. I'll do it a little later.

File: .github/workflows/rust.yml (hunk 4 above)
Comment (Mephistophiles) [reply .github/workflows/rust.yml]: Removed.

File: .github/workflows/rust.yml (hunk 3 above)
Comment (Mephistophiles) [reply .github/workflows/rust.yml:41]: Switched to rust from archlinux

File: .github/workflows/rust.yml (hunk 2 above)
Comment (Mephistophiles) [reply .github/workflows/rust.yml:31]: Needed for using archlinux instead of ubuntu (default linux container)

Comment (Morganamilo) [reply .github/workflows/rust.yml:31]: Works for me without it at https://github.com/Morganamilo/alpm.rs/blob/master/.github/workflows/alpm.yml

File: .github/workflows/rust.yml (hunk 5 above)
Comment (Morganamilo) [reply .github/workflows/rust.yml]: Also forgot to mention, isn't the container running as root? the sudos shouldn't be needed right?

**Comment by Mephistophiles** [comment]
@Morganamilo small question: do you need the build badge in the README.md?
example: Rust

**Comment by Morganamilo** [comment]
Not really, the commit already shows the ci status.

File: .github/workflows/rust.yml (hunk 5 above)
Comment (Mephistophiles) [reply .github/workflows/rust.yml]: Oh, yes. Removed.

File: .github/workflows/rust.yml (hunk 2 above)
Comment (Mephistophiles) [reply .github/workflows/rust.yml]: oh, I apologize for not immediately thinking to do so :(

File: .github/workflows/rust.yml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    steps:
+    - name: Install Packages
+      run: pacman -Syu rust clang gcc --noconfirm --needed
+    - uses: actions/checkout@v2
+      with:
+        fetch-depth: 0
+
+    - uses: actions-rs/clippy-check@v1
+      with:
+        token: ${{ secrets.GITHUB_TOKEN }}
```
Comment (Morganamilo) [inline .github/workflows/rust.yml:78]: Also why is the secret key needed for this?

Comment (Mephistophiles) [reply .github/workflows/rust.yml:78]: It's only needed for forked repos. I will remove it before change to "Ready for Review" state.

**Comment by Morganamilo** [comment]
By the way, if you're interested in doing CI at all, It would be nice to get #83 solved.  Yay does it with this https://github.com/Jguer/yay/blob/next/.github/workflows/multiarch-build.yml but I'm not really familiar with github actions or docker.

**Comment by patrickelectric** [comment]
Hey, any update ?

**Pull request closed without merging**